colexec: plumb through Allocator skeleton

yuzefovich opened the pull request:
This commit adds the plumbing of newly added Allocator which is
responsible for doling out batches to operators within a fixed budget.
The budgeting logic doesn't exist yet, but this commit adds the new type
and plumbs it through operators which use a variable amount of memory.
Next step is to actually measure the memory required when NewBatch or
Append is called and track that against a budget.

A linter rule is added that prohibits usage of coldata.NewMem* methods
within sql/colexec and sql/colflow packages.

Based on: #38394.
Fixes: #38056.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42232)
<!-- Reviewable:end -->

yuzefovich commented:
I addressed all the comments that were left on #38394.

I also made a couple of modifications (like introducing a batch allocator to non-buffering operators) which I'm not sure about: do we want to only buffering operators go through the batch allocator or do we want everything to go through it? I feel like the former.

I also think that caring only about allocating new batches and `Append`s is not enough - we need to care about `Copy`s and `Set`s as well since those can allocate memory as well (for example, when dealing with bytes). This could be left to a follow-up PR.

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt5Od5BBivMidVkE_X3:-Lt5Od5BBivMidVkE_X4:b-3qejep) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
> ```Go
> //
> // TODO(yuzefovich): Add memory budget logic.
> type BatchAllocator struct{}
> ```

how complicated is this logic going to be? if it's anything like the `mon` package (or going to use the `mon` package) it probably belongs somewhere under pkg/sql instead of in col/coldata


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 27 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @yuzefovich)

___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtUzLSoCmar71m9_CUT:-LtUzLSp998zoQNZrO6_:bkscza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
> ```Go
> //
> // TODO(yuzefovich): Add memory budget logic.
> type BatchAllocator struct{}
> ```

We could maybe just call this `Allocator`?
___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt5Od5BBivMidVkE_X3:-LtUzSsoEwcWt-UiViN2:b-eh9gvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

how complicated is this logic going to be? if it's anything like the `mon` package (or going to use the `mon` package) it probably belongs somewhere under pkg/sql instead of in col/coldata
</blockquote></details>

It's not going to be too complicated. The initial idea was to pre-reserve an amount of work memory for buffering operators from a monitor at plan-time and enforce that limit via an operator-local counter, but we might end up plumbing down a `mon` for online monitoring. I'm hesitant to put this in a different package since the idea is that only `Allocator`s will ever be used to create new batches, and it doesn't seem right to have the `Batch` allocation API in a different package, but happy to hear other options.
___
*[pkg/col/coldata/batch_allocator.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV3BuqAaRqxvjkRP7F:-LtV3BuqAaRqxvjkRP7G:b-h473n) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L32)):*
> ```Go
> // size.
> func (*BatchAllocator) NewMemBatchWithSize(types []coltypes.T, size int) (batch Batch, err error) {
> 	return NewMemBatchWithSize(types, size), nil
> ```

We could maybe also go the extra step of making these package-global functions private. This would require changing streaming operators as well (probably) but it's something we want to do at some point anyway.
___
*[pkg/sql/colexec/aggregator.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV3rrm16zK0A3pW290:-LtV3rrm16zK0A3pW291:b-29m0of) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/sql/colexec/aggregator.go#L90)):*
> ```Go
> 	OneInputNode
> 
> 	allocator coldata.BatchAllocator
> ```

Should these be pointers?
___
*[pkg/sql/colexec/coalescer.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV4FJyEdTiWHLnA50e:-LtV4FJyEdTiWHLnA50f:butzp6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/sql/colexec/coalescer.go#L22)):*
> ```Go
> 
> // TODO(yuzefovich): we actually seem to never use coalescerOp. Think through
> // whether we want to plan it in some cases.
> ```

I would probably delete it. One option would be to extend the `deselectorOp` with coalescing behavior, but I'm not sure we want that (e.g. consider an outbox with a deselector op on top, we probably want to push the first batch of data ASAP)


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt5Od5BBivMidVkE_X3:-LtVCGIbDYO4ma6lqrj6:bhlh3hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It's not going to be too complicated. The initial idea was to pre-reserve an amount of work memory for buffering operators from a monitor at plan-time and enforce that limit via an operator-local counter, but we might end up plumbing down a `mon` for online monitoring. I'm hesitant to put this in a different package since the idea is that only `Allocator`s will ever be used to create new batches, and it doesn't seem right to have the `Batch` allocation API in a different package, but happy to hear other options.
</blockquote></details>

My instinct here is that this will be too specific to sql execution's needs and not right for other users of coldata (but it's hard to say without seeing more detail than is present in this PR). There are already a couple uses of coldata inside crdb now, and I actually think of everything under pkg/col as general purpose enough that you could use it from another go project. It would be nice to keep it that way. If we put an Allocator in here, we should commit to the logic being general purpose and not just what exec needs.

My counterproposal is to put the allocator under pkg/sql somewhere and add a lint rule that disallows use of coldata.NewBatch and coldata.NewBatchWithSize from inside pkg/sql. We already have similar lint rules that disallow use of time.Now in favor of timeutil.Now (among other examples) so this is likely straightforward to add.

That will allow you maximum flexibility to make the Allocator whatever exec needs without worrying about whether that's appropriate for a more general coldata library. If at the end, we've ended up with a nice general purpose Allocator (or something close enough to that that it's worth a small cleanup), then we can always move it here.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @yuzefovich)

___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt5Od5BBivMidVkE_X3:-LtVIwzh4hlM6IYIFy7n:bp0rksw) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

My instinct here is that this will be too specific to sql execution's needs and not right for other users of coldata (but it's hard to say without seeing more detail than is present in this PR). There are already a couple uses of coldata inside crdb now, and I actually think of everything under pkg/col as general purpose enough that you could use it from another go project. It would be nice to keep it that way. If we put an Allocator in here, we should commit to the logic being general purpose and not just what exec needs.

My counterproposal is to put the allocator under pkg/sql somewhere and add a lint rule that disallows use of coldata.NewBatch and coldata.NewBatchWithSize from inside pkg/sql. We already have similar lint rules that disallow use of time.Now in favor of timeutil.Now (among other examples) so this is likely straightforward to add.

That will allow you maximum flexibility to make the Allocator whatever exec needs without worrying about whether that's appropriate for a more general coldata library. If at the end, we've ended up with a nice general purpose Allocator (or something close enough to that that it's worth a small cleanup), then we can always move it here.
</blockquote></details>

Fair enough, thanks for your input! We'd probably want to disallow all uses of `coldata.{NewBatch,NewBatchWithSize}` in `colexec` as well. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-:-Lt_rrVx1cc4r7vkQ6Cq:b-828fna)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @danhhz)

___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt5Od5BBivMidVkE_X3:-LtX38U99IQ6ppzz2tdY:b-p3u6dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Fair enough, thanks for your input! We'd probably want to disallow all uses of `coldata.{NewBatch,NewBatchWithSize}` in `colexec` as well. 
</blockquote></details>

Thanks for the suggestion! I like the idea of placing it into `sql/colexec` and enforcing its usage by a linter. Done.
___
*[pkg/col/coldata/batch_allocator.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtUzLSoCmar71m9_CUT:-Lt_mhLS64jmuXZ7DyWI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L22)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We could maybe just call this `Allocator`?
</blockquote></details>

Done.
___
*[pkg/col/coldata/batch_allocator.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV3BuqAaRqxvjkRP7F:-Lt_n_bG3pixqlwbsmc_:b-p3xn3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/col/coldata/batch_allocator.go#L32)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We could maybe also go the extra step of making these package-global functions private. This would require changing streaming operators as well (probably) but it's something we want to do at some point anyway.
</blockquote></details>

I think the linter addresses that. In order to not make any exceptions, I plumbed the allocator everywhere.
___
*[pkg/sql/colexec/aggregator.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV3rrm16zK0A3pW290:-LtX1j449f90UveR7q9n:b3obbms) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/sql/colexec/aggregator.go#L90)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should these be pointers?
</blockquote></details>

Yeah, I think so, thanks.
___
*[pkg/sql/colexec/allocator.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt_sTHS0Wm7KFEwPw-o:-Lt_sTHS0Wm7KFEwPw-p:bi5m812) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f3a368ed50a844075b8acbc775a937e0158fbd7/pkg/sql/colexec/allocator.go#L33)):*
> ```Go
> 
> // TODO(yuzefovich): add NewMemColumn method.
> // TODO(yuzefovich): it is possible for a batch obtained from an Allocator to
> ```

Currently only (among relevant to the vectorized engines files)`ArrowToBatch` method of `ArrowBatchConverter` can `Reset` the batch which could allocate a new batch but shouldn't because in the `Inbox` we allocate one of `coldata.BatchSize` size. I'm not sure whether there is anything to do. Maybe we should prohibit usage of `Reset` method in `colexec` and `colflow`?
___
*[pkg/sql/colexec/coalescer.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-LtV4FJyEdTiWHLnA50e:-Lt_nuz14viVZGUGQeqQ:b-6yyxyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/39513f961f56ab5b0f2c9e1bcf2c4d571167d934/pkg/sql/colexec/coalescer.go#L22)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would probably delete it. One option would be to extend the `deselectorOp` with coalescing behavior, but I'm not sure we want that (e.g. consider an outbox with a deselector op on top, we probably want to push the first batch of data ASAP)
</blockquote></details>

I agree that we should just delete it. I think that `coalescerOp` is not used because other operators (like joins) have the coalescing behavior inside of them. I'll remove it in a separate PR.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I wonder if it would be better to `panic` with a `VectorizedInternalPanic` instead of returning an error?

> Reviewed 68 of 68 files at r2, 1 of 1 files at r3, 10 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-:-LtaAj5rDCwjmiPWkWbR:blho5qd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @danhhz, and @yuzefovich)

___
*[pkg/sql/colexec/allocator.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt_sTHS0Wm7KFEwPw-o:-Lta4yzW4GReHJBYm0Oq:b-jg1lrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f3a368ed50a844075b8acbc775a937e0158fbd7/pkg/sql/colexec/allocator.go#L33)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Currently only (among relevant to the vectorized engines files)`ArrowToBatch` method of `ArrowBatchConverter` can `Reset` the batch which could allocate a new batch but shouldn't because in the `Inbox` we allocate one of `coldata.BatchSize` size. I'm not sure whether there is anything to do. Maybe we should prohibit usage of `Reset` method in `colexec` and `colflow`?
</blockquote></details>

Hmmm, yeah. I don't think we'd ever want to use `Reset` instead of `ResetInternalBatch`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I prefer returning an error in this case so that the caller could decide what to do with it. If we're panicking, then all buffering operators that know how to spill to disk would need to catch panics and check whether it was an OOM error, but if we return an error, there is no need for catching panics. If an operator gets an error and doesn't know how to spill, it will then panic itself.

What is your thinking for panicking here?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-:-LtaDgoT0beu5JMGuP_e:brobq7z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @danhhz)

___
*[pkg/sql/colexec/allocator.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt_sTHS0Wm7KFEwPw-o:-LtaGPUj2Yx9tnEPYX-5:b-q9y2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f3a368ed50a844075b8acbc775a937e0158fbd7/pkg/sql/colexec/allocator.go#L33)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Hmmm, yeah. I don't think we'd ever want to use `Reset` instead of `ResetInternalBatch`.
</blockquote></details>

I agree. Do you know if there is a way to distinguish different `Reset` methods in a linter? I feel like "no", but we use different Resets in colexec and colflow, and I don't have a better idea other than whitelisting all of them with a line number which can be burdensome and too verbose (right now there are like 12 calls to `Reset()` methods which are different from `Batch.Reset()`).


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
My thinking is that buffering operators won't be the ones to actually handle the error. Taking a sort for example, I was thinking that we'd have a separate external sort operator that has the in-memory sort as input and essentially just normally forwards batches returned, but catches any panics. If the panic is an out of memory, it would store all the data stored in the in-memory sorter to disk, and proceed with the rest of the external sort. This makes it so that any operator that doesn't care never has to handle any sort of errors from allocations. This is part of what I meant in https://github.com/cockroachdb/cockroach/issues/42408

> Reviewed 1 of 13 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-:-LtaV4v-D2S7Guoxw5zo:b3hbolv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/allocator.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt_sTHS0Wm7KFEwPw-o:-LtaUxCKDMOCa2g0eiix:bfguzzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f3a368ed50a844075b8acbc775a937e0158fbd7/pkg/sql/colexec/allocator.go#L33)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I agree. Do you know if there is a way to distinguish different `Reset` methods in a linter? I feel like "no", but we use different Resets in colexec and colflow, and I don't have a better idea other than whitelisting all of them with a line number which can be burdensome and too verbose (right now there are like 12 calls to `Reset()` methods which are different from `Batch.Reset()`).
</blockquote></details>

I don't think so. I think it would be fine to just put a warning that `ResetInternalBatch` could allocate.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: plumb through Allocator skeleton

This commit adds the plumbing of newly added Allocator which is
responsible for doling out batches to operators within a fixed budget.
The budgeting logic doesn't exist yet, but this commit adds the new type
and plumbs it through operators which use a variable amount of memory.
Next step is to actually measure the memory required when NewBatch or
Append is called and track that against a budget.

A linter rule is added that prohibits usage of coldata.NewMem* methods
within sql/colexec and sql/colflow packages.

Release note: None

yuzefovich reviewed (commented):
Ok, I see your point. Removed the error from the return parameters.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-:-LtajRpJ0WHtJ1dIacYB:b-xl0qgs)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/allocator.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42232#-Lt_sTHS0Wm7KFEwPw-o:-LtajfkcCp5HfrTPrUY5:bbn0f79) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f3a368ed50a844075b8acbc775a937e0158fbd7/pkg/sql/colexec/allocator.go#L33)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think so. I think it would be fine to just put a warning that `ResetInternalBatch` could allocate.
</blockquote></details>

Yeah, I added a comment to call it out.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
